### PR TITLE
Bump ESLint and related dependencies

### LINKED ISSUE #2973: Migrate to ESLint flat config (and v9)
The major version of ESLint is now v9 which includes a couple of breaking changes, most notably:

* No more `.eslintrc`, moving to `eslint.config.js`
* Style-related rules are now moved to `@stylistic/eslint-plugin`
* `@typescript-eslint/***` packages are now `typescript-eslint`

Will have to investigate if this can cause issues due to conflicts with CRA's ESLint version.

May or may not be blocked by Vite migration.

Note that VSCode support is still flagged as "experimental" and we'll need to commit `.vscode/settings.json` into the repository for this.

### PR DESCRIPTION
**Pull request by RichDom2185**
### Description

Closes #2973. Refer to the linked issue for details.

`eslint-plugin-react-refresh` is a newly installed plugin that will ease our migration to Vite. We install the canary version as the stable version that supports ESLint v9 will only be released alongside React 19.

Also added TODOs to eventually make the code quality stricter.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] Code quality improvements

### How to test

`yarn eslint` works.

### Checklist

- [x] I have tested this code
- [ ] I have updated the documentation

### CODE REVIEW

File: eslint.config.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  ...tseslint.configs.recommended,
+  // TODO: Enable when ready
+  // {
+  //   plugins: {
+  //     'react-refresh': reactRefresh
+  //   },
+  //   rules: {
+  //     'react-refresh/only-export-components': 'warn'
+  //   }
+  // },
```
Comment (RichDom2185) [inline eslint.config.js:25]: Should I resolve the warnings in this PR @chownces? Feels a bit out of scope though, I think better to resolve all warnings in one go in a separate PR.

**Comment by RichDom2185** [comment]
As expected, CRA makes the dev experience a nightmare&hellip;

image

It's using an old version of ESLint which makes it incompatible. I think we should disable ESLint in the dev server and let us handle the linting separately.

**Review by chownces — APPROVED** [review]
Can we track these pending comments as new issues with the high/ critical tag? I assume these changes will be done during the Vite or React19 migration.

File: package.json [hunk 2]
```diff
@@ -132,15 +132,17 @@
...
     "babel-jest": "^29.7.0",
     "buffer": "^6.0.3",
     "canvas": "^2.11.2",
     "constants-browserify": "^1.0.0",
     "coveralls": "^3.1.1",
     "cross-env": "^7.0.3",
-    "eslint": "^8.57.0",
+    "eslint": "^9.1.1",
+    "eslint-plugin-react": "^7.34.1",
+    "//": "See: https://github.com/facebook/react/issues/28313#issuecomment-2076798972, https://github.com/t3-oss/create-t3-turbo/issues/984#issuecomment-2076413457",
```
Comment (chownces) [inline package.json:143]: Let's remember to clean this up post migration

Comment (RichDom2185) [reply package.json:144]: Done in #2990

File: eslint.config.js (hunk 1 above)
Comment (RichDom2185) [reply eslint.config.js:25]: Done in #2990
